lib/api/src/module.rs

```diff
@@ -411,14 +414,40 @@ impl Module {
         &self.artifact.module_ref()
     }
 
-    /// Gets the [`Artifact`] used internally by the Module.
+    /// Query the artifact to fetch information from it without
+    /// extracting the artifact from the module.
+    ///
+    /// The artifact defined in the `Module` is abstracted behind `dyn
+    /// Artifact`. When querying, downcasting to a concrete artifact
+    /// type is done by inferring the type from the closure used to
+    /// query. Something like this:
+    ///
+    /// ```rust,no_run
+    /// use wasmer_engine_object_file::ObjectFileArtifact;
+    ///
+    /// let symbol_registry = unsafe {
+    ///     module.query_artifact(|artifact: &ObjectFileArtifact| {
+    ///         Some(Arc::clone(artifact.symbol_registry()))
+    ///     })
+    /// };
+    /// ```
+    ///
+    /// If downcasting the artifact to `A` fails, the method will
+    /// return `None`.
     ///
     /// This API is hidden because it's not necessarily stable;
     /// this functionality is required for some core functionality though, like
-    /// the object file engine.
+    /// the [`wasmer-engine-object-file`].
     #[doc(hidden)]
-    pub fn artifact(&self) -> &Arc<dyn Artifact> {
-        &self.artifact
+    pub unsafe fn query_artifact<A, Q, R>(&self, query: Q) -> Option<Arc<R>>
```

Comment:
`unsafe` functions must have a `# Safety` section explaining how the function must be used to avoid UB, etc.

Reply:
I think we can remove this one altogether. No need to return the artifact at all.